cl/citations/__init__.py

```diff
@@ -0,0 +1,147 @@
+        % (span_class, data_attr, self.antecedent_guess, inner_html)
+    )
+
+
+def supra_citation_as_html(self):
+    inner_html = (
+        '<span class="antecedent_guess">%s</span>' % self.antecedent_guess
+    )
+    if self.volume:
+        inner_html += (
+            '\\g<1><span class="volume">%d</span>\\g<2>supra' % self.volume
+        )
+        if self.page:
+            inner_html += (
+                ',\\g<3>at\\g<4><span class="page">%s</span>\\g<5>' % self.page
+            )
+        else:
+            inner_html += "\\g<3>"
+    else:
+        inner_html += "\\g<1>supra"
+        if self.page:
+            inner_html += (
+                ',\\g<2>at\\g<3><span class="page">%s</span>\\g<4>' % self.page
+            )
+        else:
+            inner_html += "\\g<2>"
+
+    span_class = "citation"
+    if self.match_url:
+        inner_html = '<a href="%s">%s</a>' % (self.match_url, inner_html)
+        data_attr = ' data-id="%s"' % self.match_id
+    else:
+        span_class += " no-link"
+        data_attr = ""
+    return '<span class="%s"%s>%s</span>' % (
+        span_class,
+        data_attr,
+        inner_html,
+    )
+
+
+def id_citation_as_html(self):
+    def generate_after_token_html():
+        # First, insert the regex backreferences between each "after token"
+        # The group numbers of each backreference (g<NUMBER>) must be
+        #   dynamically generated because total number of "after tokens" varies
+        # Produces something like this:
+        # "\\g<2>after_token_1\\g<3>after_token_2\\g<4>after_token_3" ...
+        template = "\\g<%s>%s"
+        after_token_html = "".join(
+            [
+                template % (str(i + 2), t)
+                for i, t in enumerate(self.after_tokens)
+            ]
+        )
+
+        # Then, append one final backreference to the end of the string
+        after_token_html += "\\g<" + str(len(self.after_tokens) + 2) + ">"
+
+        # Return the full string
+        return after_token_html
+
+    span_class = "citation"
+    after_token_html = generate_after_token_html()
+    if self.match_url:
+        if self.has_page:
+            id_string_template = (
+                '<a href="%s"><span class="id_token">%s</span>%s</a>'
+            )
+        else:
+            id_string_template = (
+                '<a href="%s"><span class="id_token">%s</span></a>%s'
+            )
+        id_string = id_string_template % (
+            self.match_url,
+            self.id_token,
+            after_token_html,
+        )
+        data_attr = ' data-id="%s"' % self.match_id
+    else:
+        id_string = '<span class="id_token">%s</span>%s' % (
+            self.id_token,
+            after_token_html,
+        )
+        span_class += " no-link"
+        data_attr = ""
+    return '<span class="%s"%s>\\g<1>%s</span>' % (
+        span_class,
+        data_attr,
+        id_string,
+    )
+
+
+FullCitation.as_html = full_citation_as_html
```

Comment:
This feels like a funny way to do this. Why not subclass FullCitation et al and add the as_html function that way? Then you can do the imports the same way, but it'll be more clear what particular functionality you're adding to the classes.

Should I look at all this HTML stuff in these functions? I don't recognize it, but if it's just copy/paste from earlier code, then I'll trust you got it right.

Reply:
I agree that it is a little awkward, but I don't entirely understand your solution. `eyecite.find_citations.get_citations()` is going to return its version of `eyecite.models.FullCitation` etc. So, yes, we could subclass that model in CL (something like `class CLFullCitation(FullCitation):`), but that isn't the object that will be returned by `get_citations()`. Hence, we need to retroactively graft our extra methods onto the `eyecite.models.FullCitation` model _itself_, because it is objects of that class that we are passing around. At least, I think so?

And yeah, the body of the functions are just copy-pasted.